Kernel/Thread.cpp

```diff
@@ -45,6 +45,14 @@
 
 namespace Kernel {
 
+SpinLock<u8> Thread::g_tid_map_lock;
+HashMap<pid_t, Thread*>* Thread::g_tid_map;
```

Comment:
Seems like we should use `ThreadID` instead of `pid_t` here.

Reply:
I tried that but I couldn't... No `hash` function it seems, though I don't quite understand the type magic for those...

Reply:
Maybe there's a way to implement a template `hash` function for those types?

Reply:
@BenWiederhake Can we add some templated `hash` function to `DistinctNumeric` so that we can use things like `ThreadID` as keys in `HashMap`?

Reply:
Tried to add something like this to DistinctNumeric.h but not sure what the syntax has to be like exactly:
```c++
template<typename T, typename... Args>
struct Traits<DistinctNumeric<T, Args...>> : public Traits<T> {};
```

Reply:
I also took a look at this, the core problem is that a parameter pack declared with 'auto' will only accept NTTP's, and a parameter pack declared with 'typename' (or 'class') will only accept types.
taking an extra `typename X` at the end of that specialisation will somehow confuse the compiler, and make it unable to infer _any_ of the types in the template params, so that's out too.

The only way I can think of is to just move the last parameter of `DistinctNumeric` to the second parameter position - this will make reading the errors a little bit more difficult, but traits are more useful imho.

Reply:
@alimpfard thanks for looking into this! Do you mind if I cherry pick your change into this PR? That way @awesomekling can just merge this PR without merge conflicts (we had that issue yesterday when when I had one PR be based on another).